utils/tfhe-versionable-derive/src/dispatch_type.rs

```diff
@@ -0,0 +1,330 @@
+                let dest = parse_quote! { #dest_type };
+                let constructor = self.generate_conversion_constructor_owned("value")?;
+
+                let assoc_to_orig = generate_conversion_impl(
+                    &src,
+                    &dest,
+                    &impl_generics,
+                    where_clause,
+                    &constructor,
+                    "value",
+                )?;
+
+                // Wraps the highest version into the dispatch enum
+                let src_type = self.latest_version_type()?;
+                let src = parse_quote! { &#src_type };
+                let dest_ident = self.ident();
+                let dest = parse_quote! { #dest_ident #ty_generics };
+                let constructor = self.generate_conversion_constructor_ref("value")?;
+
+                let orig_to_assoc = generate_conversion_impl(
+                    &src,
+                    &dest,
+                    &impl_generics,
+                    where_clause,
+                    &constructor,
+                    "value",
+                )?;
+
+                Ok(vec![orig_to_assoc, assoc_to_orig])
+            }
+        }
+    }
+
+    fn ident(&self) -> Ident {
+        match &self.kind {
+            AssociatedTypeKind::Ref(_) => {
+                format_ident!("{}Dispatch", self.orig_type.ident)
+            }
+            AssociatedTypeKind::Owned => {
+                format_ident!("{}DispatchOwned", self.orig_type.ident)
+            }
+        }
+    }
+
+    fn lifetime(&self) -> Option<&Lifetime> {
+        match &self.kind {
+            AssociatedTypeKind::Ref(lifetime) => lifetime.as_ref(),
+            AssociatedTypeKind::Owned => None,
+        }
+    }
+
+    fn as_trait_param(&self) -> Option<syn::Result<&Type>> {
+        Some(self.latest_version_type())
+    }
+
+    fn inner_types(&self) -> syn::Result<Vec<&Type>> {
+        self.version_types()
+    }
+}
+
+impl DispatchType {
+    /// Returns the error sent to the user for a wrong use of this macro
+    fn error(&self) -> syn::Error {
+        syn::Error::new(
+            self.orig_type.span(),
+            "VersionsDispatch should be used on a enum with single anonymous field variants",
+        )
+    }
+
+    fn generics(&self) -> syn::Result<Generics> {
+        let mut generics = self.orig_type.generics.clone();
+        if let AssociatedTypeKind::Ref(Some(lifetime)) = &self.kind {
+            add_lifetime_bound(&mut generics, lifetime);
+        }
+
+        add_trait_where_clause(&mut generics, self.inner_types()?, &[VERSION_TRAIT_NAME])?;
+
+        add_trait_bound(&mut generics, VERSIONIZE_TRAIT_NAME)?;
+
+        Ok(generics)
+    }
+
+    /// Returns the number of versions in this dispatch enum
+    fn versions_count(&self) -> usize {
+        self.orig_type.variants.len()
+    }
+
+    /// Returns the latest version of the original type, which is the last variant in the enum
```

Comment:
As of today the macro does not enforce anything for the name of the variants in the dispatch enum. In the examples I chose `V0, V1, ...` but you could also name them `A, B, C`, `Version0, Version1,...` or anything you want. So it is not possible to check that they are correctly ordered. Should I require a specific naming scheme ?

Reply:
I don't know if we want to be strict and enforce a naming scheme, even if this system is only for us it might create some restriction, but on the other end, it might help us avoid errors :thinking: 

Maybe the naming scheme could be an optional argument via attribute (i.e no attribute = no naming scheme enforced, `#[Versionize(naming=V$n)]`, `#[Versionize(naming=Version$n)]` but that's probabling going to far :cold_face: )

Reply:
The conversion traits were raising errors if fields were not ordered properly IIRC

I don’t think it’s necessary to force the naming scheme

Reply:
Ok so I will leave it like this for the moment.